File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/fragments/investigation/evidence/views/ghost/GhostView.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.evidence.views.ghost;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;
import android.os.Bundle;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.LinearLayout;

import androidx.annotation.IntegerRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.appcompat.widget.LinearLayoutCompat;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.view.GestureDetectorCompat;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.evidence.data.InvestigationData;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.evidence.data.investigationtype.Ghost;
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.EvidenceViewModel;

public abstract class GhostView extends ConstraintLayout {

    private EvidenceViewModel evidenceViewModel;

    private Ghost ghostData;

    private @IntegerRes int neutralSelColor, negativeSelColor, positiveSelColor;

    public GhostView(@NonNull Context context) {
        super(context);
        initView();
    }

    public GhostView(@NonNull Context context, @Nullable @org.jetbrains.annotations.Nullable AttributeSet attrs) {
        super(context, attrs);
        initView();
    }

    public GhostView(@NonNull Context context, @Nullable @org.jetbrains.annotations.Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView();
    }

    public GhostView(@NonNull Context context, @Nullable @org.jetbrains.annotations.Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initView();
    }

    public void initView() {
        inflate(getContext(), R.layout.item_investigation_ghost, this);

        LinearLayoutCompat.LayoutParams params =
                new LinearLayoutCompat.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT,
                        ConstraintLayout.LayoutParams.WRAP_CONTENT, 1f);
        setLayoutParams(params);

        TypedValue typedValue = new TypedValue();
        Resources.Theme theme = getContext().getTheme();
        theme.resolveAttribute(R.attr.neutralSelColor, typedValue, true);
        neutralSelColor = typedValue.data;
        theme.resolveAttribute(R.attr.negativeSelColor, typedValue, true);
        negativeSelColor = typedValue.data;
        theme.resolveAttribute(R.attr.positiveSelColor, typedValue, true);
        positiveSelColor = typedValue.data;

    }

    @SuppressLint("ClickableViewAccessibility")
    public void build(EvidenceViewModel evidenceViewModel, int groupIndex) {

        this.evidenceViewModel = evidenceViewModel;
        this.ghostData = evidenceViewModel.getInvestigationData().getGhost(groupIndex);

        AppCompatTextView nameView = findViewById(R.id.label_name);
        LinearLayoutCompat linearLayout_iconRow = findViewById(R.id.icon_container);

        String ghostName = ghostData.getName();
        nameView.setText(ghostName);

        redrawGhostRejectionStatus(ghostData, groupIndex, false);

        if(linearLayout_iconRow != null) {

            int k = 0;
            for (; k < ghostData.getEvidence().length; k++) {

                AppCompatImageView evidenceIcon =
                        (AppCompatImageView) linearLayout_iconRow.getChildAt(k);
                evidenceIcon.setImageResource(ghostData.getEvidence()[k].getIcon());

                switch (ghostData.getEvidence()[k].getRuling()) {
                    case POSITIVE -> evidenceIcon.setColorFilter(positiveSelColor);
                    case NEGATIVE -> evidenceIcon.setColorFilter(negativeSelColor);
                    case NEUTRAL -> evidenceIcon.setColorFilter(neutralSelColor);
                }

            }

            for(; k < linearLayout_iconRow.getChildCount(); k ++) {
                linearLayout_iconRow.getChildAt(k).setVisibility(View.GONE);
            }
        }

        setVisibility(View.INVISIBLE);
        setAlpha(0);
        setId(groupIndex);


        GestureDetectorCompat swipeListener = new GestureDetectorCompat(getContext(),
                new GhostSwipeListener(groupIndex)
        );

        nameView.setOnTouchListener((v, motionEvent) -> {
            swipeListener.onTouchEvent(motionEvent);
            return true;
        });

        addOnLayoutChangeListener((thisGhostView, i, i1, i2, i3, i4, i5, i6, i7) -> {

            redrawGhostRejectionStatus(ghostData, groupIndex, false);

            if(linearLayout_iconRow != null) {
                for (int l = 0; l < ghostData.getEvidence().length; l++) {

                    AppCompatImageView evidenceIcon = linearLayout_iconRow.getChildAt(l).findViewById(R.id.evidence_icon);

                    switch (ghostData.getEvidence()[l].getRuling()) {
                        case POSITIVE -> evidenceIcon.setColorFilter(positiveSelColor);
                        case NEGATIVE -> evidenceIcon.setColorFilter(negativeSelColor);
                        case NEUTRAL -> evidenceIcon.setColorFilter(neutralSelColor);
                    }
                }
            }
        });

        post(new Runnable() {
            @Override
            public void run() {
                animate()
                        .setListener(new AnimatorListenerAdapter() {
                            @Override
                            public void onAnimationStart(Animator animation) {
                                super.onAnimationStart(animation);
                                setVisibility(View.VISIBLE);
                            }}
                        ).alpha(1).setStartDelay((long)(10f * groupIndex)).setDuration(100);
            }
        });
    }

    public void forceUpdateComponents() {
        LinearLayoutCompat linearLayout_iconRow = findViewById(R.id.icon_container);
        if(linearLayout_iconRow == null) { return; }

        for (int k = 0; k < ghostData.getEvidence().length; k++) {

            AppCompatImageView evidenceIcon =
                    (AppCompatImageView) linearLayout_iconRow.getChildAt(k);

            switch (ghostData.getEvidence()[k].getRuling()) {
                case POSITIVE -> evidenceIcon.setColorFilter(positiveSelColor);
                case NEGATIVE -> evidenceIcon.setColorFilter(negativeSelColor);
                case NEUTRAL -> evidenceIcon.setColorFilter(neutralSelColor);
            }
        }

        redrawGhostRejectionStatus(ghostData, ghostData.getId(), true);

    }

    private void redrawGhostRejectionStatus(Ghost ghost, int index, boolean animate) {

        int score = ghost.getEvidenceScore();
        AppCompatImageView statusIcon = findViewById(R.id.icon_status);

        boolean rejectionStatus = evidenceViewModel.getRejectionPile()[index];
        if (rejectionStatus) {
            statusIcon.setImageLevel(1);

        } else if (score <= -5) {
            statusIcon.setImageLevel(2+ (int) (Math.random() * 3));

        } else if (score == 3) {
            statusIcon.setImageLevel(5);

        } else {
            statusIcon.setImageLevel(0);
        }

    }

    public class GhostSwipeListener extends GestureDetector.SimpleOnGestureListener {

        private final int index;

        public GhostSwipeListener(
                int index) {
            super();

            this.index = index;
        }

        @Override
        public boolean onFling(MotionEvent event1, MotionEvent event2,
                               float velocityX, float velocityY) {

            boolean status = !evidenceViewModel.swapStatusInRejectedPile(index);

            evidenceViewModel.getGhostOrderData().updateOrder();

            redrawGhostRejectionStatus(evidenceViewModel.getInvestigationData().getGhost(index), index, true);

            Bundle params = new Bundle();
            params.putString("event_type", "ghost_swiped");
            params.putString("event_details", status ? "ghost_impartial" : "ghost_rejected");
            //analytics.logEvent("event_investigation", params);

            return true;
        }

        @Override
        public boolean onSingleTapUp(@NonNull MotionEvent e) {
            createPopup();

            return super.onSingleTapConfirmed(e);
        }

    }

    public abstract void createPopup();
}

File: app-view/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/activities/investigation/evidence/views/investigation/section/ghost/GhostView.kt
package com.tritiumgaming.phasmophobiaevidencepicker.activities.investigation.evidence.views.investigation.section.ghost

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.View
import android.widget.LinearLayout
import androidx.annotation.IntegerRes
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.appcompat.widget.LinearLayoutCompat
import androidx.constraintlayout.widget.ConstraintLayout
import com.tritiumgaming.phasmophobiaevidencepicker.R
import com.tritiumgaming.phasmophobiaevidencepicker.data.viewmodels.InvestigationViewModel
import com.tritiumgaming.phasmophobiaevidencepicker.data.viewmodels.models.investigation.investigationmodels.investigationtype.evidence.EvidenceModel.Ruling
import com.tritiumgaming.phasmophobiaevidencepicker.data.viewmodels.models.investigation.investigationmodels.investigationtype.ghost.GhostModel
import com.tritiumgaming.phasmophobiaevidencepicker.utils.ColorUtils.getColorFromAttribute

class GhostView : ConstraintLayout {

    private var investigationViewModel: InvestigationViewModel? = null
    private var ghostModel: GhostModel? = null

    @IntegerRes private var neutralSelColor = 0
    @IntegerRes private var negativeSelColor = 0
    @IntegerRes private var positiveSelColor = 0

    constructor(context: Context) :
            super(context)

    constructor(context: Context, attrs: AttributeSet?) :
            super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) :
            super(context, attrs, defStyleAttr, defStyleRes)

    init {
        inflate(context, R.layout.item_investigation_ghost, this)
        setDefaults()
    }

    private fun setDefaults() {
        layoutParams =
            LinearLayoutCompat.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LayoutParams.WRAP_CONTENT, 1f
            )

        neutralSelColor = getColorFromAttribute(context, R.attr.neutralSelColor)
        negativeSelColor = getColorFromAttribute(context, R.attr.negativeSelColor)
        positiveSelColor = getColorFromAttribute(context, R.attr.positiveSelColor)
    }

    @SuppressLint("ClickableViewAccessibility")
    fun build(investigationViewModel: InvestigationViewModel, groupIndex: Int) {
        this.investigationViewModel = investigationViewModel
        this.ghostModel = investigationViewModel.investigationModel?.ghostListModel?.getAt(groupIndex)

        val nameView = findViewById<AppCompatTextView>(R.id.label_name)
        val iconRowLayout = findViewById<LinearLayoutCompat>(R.id.icon_container)

        nameView.text = ghostModel?.name?.let { resId -> context.getString(resId) }

        ghostModel?.let { redrawGhostRejectionStatus(it, groupIndex, false) }

        if (iconRowLayout != null) {
            var k = 0
            while (k < (ghostModel?.evidence?.size ?: 0)) {
                val currentEvidence = ghostModel?.evidence?.get(k) ?: return

                val evidenceIcon =
                    iconRowLayout.getChildAt(k) as AppCompatImageView

                evidenceIcon.setImageResource(currentEvidence.icon)

                when (currentEvidence.ruling) {
                    Ruling.POSITIVE -> evidenceIcon.setColorFilter(positiveSelColor)
                    Ruling.NEGATIVE -> evidenceIcon.setColorFilter(negativeSelColor)
                    Ruling.NEUTRAL -> evidenceIcon.setColorFilter(neutralSelColor)
                }
                k++
            }

            while (k < iconRowLayout.childCount) {
                iconRowLayout.getChildAt(k).visibility = GONE
                k++
            }
        }

        visibility = INVISIBLE
        alpha = 0f
        id = groupIndex

        val swipeListener = GestureDetector(context, OnSwipeListener(groupIndex))
        nameView.setOnTouchListener { _: View?, motionEvent: MotionEvent? ->
            motionEvent?.let { swipeListener.onTouchEvent(motionEvent) }
            true
        }

        addOnLayoutChangeListener { _: View?, _: Int, _: Int, _: Int, _: Int,
                                    _: Int, _: Int, _: Int, _: Int ->
            ghostModel?.let { redrawGhostRejectionStatus(it, groupIndex, false) }
            if (iconRowLayout != null) {
                ghostModel?.evidence?.forEachIndexed { index, evidenceModel ->
                    val evidenceIcon = iconRowLayout.getChildAt(index)
                        .findViewById<AppCompatImageView>(R.id.evidence_icon)
                    evidenceModel?.let {
                        evidenceIcon.setColorFilter(getRulingColor(it.ruling))
                    }
                }
            }
        }

        post {
            animate()
                .setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationStart(animation: Animator) {
                        super.onAnimationStart(animation)
                        visibility = VISIBLE
                    }
                }
                ).alpha(1f).setStartDelay((10f * groupIndex).toLong()).setDuration(100)
        }
    }

    fun forceUpdateComponents() {
        val iconRowLayout = findViewById<LinearLayoutCompat>(R.id.icon_container) ?: return

        ghostModel?.let { ghostData ->
            ghostData.evidence.forEachIndexed { index, evidence ->
                val evidenceIcon =
                    iconRowLayout.getChildAt(index) as AppCompatImageView

                evidence?.ruling?.let { ruling ->
                    evidenceIcon.setColorFilter(getRulingColor(ruling))
                    getRulingColor(ruling)
                }
            }
            redrawGhostRejectionStatus(ghostData, ghostData.id, true)
        }
    }

    @IntegerRes
    private fun getRulingColor(ruling: Ruling): Int {
        return when (ruling) {
            Ruling.POSITIVE -> positiveSelColor
            Ruling.NEGATIVE -> negativeSelColor
            Ruling.NEUTRAL -> neutralSelColor
        }
    }

    private fun redrawGhostRejectionStatus(ghost: GhostModel, index: Int, animate: Boolean) {
        val statusIcon = findViewById<AppCompatImageView>(R.id.icon_status)

        investigationViewModel?.investigationModel?.getRejectionPile()?.get(index)?.let { rejectionStatus ->
            if (rejectionStatus) { statusIcon.setImageLevel(1) }
            else {
                val score = ghost.evidenceScore
                when {
                    score <= -5 -> statusIcon.setImageLevel(2 + (Math.random() * 3).toInt())
                    score == 3 -> statusIcon.setImageLevel(5)
                    else -> statusIcon.setImageLevel(0)
                }
            }
        }
    }

    inner class OnSwipeListener(private val index: Int) : SimpleOnGestureListener() {
        override fun onFling(
            e1: MotionEvent?, e2: MotionEvent, velX: Float, velY: Float): Boolean {
            investigationViewModel?.let { investigationViewModel ->
                investigationViewModel.investigationModel?.swapStatusInRejectedPile(index)
                investigationViewModel.investigationModel?.ghostOrderModel?.updateOrder()
                investigationViewModel.investigationModel?.ghostListModel?.getAt(index)?.let { ghostModel ->
                    redrawGhostRejectionStatus(ghostModel, index, true)
                }
            }
            return true
        }

        override fun onSingleTapUp(e: MotionEvent): Boolean {
            ghostViewListener?.onCreatePopup()
            return super.onSingleTapConfirmed(e)
        }
    }

    var ghostViewListener: GhostViewListener? = null
    abstract class GhostViewListener {
        abstract fun onCreatePopup()
    }
}
